airbyte-integrations/connectors/source-mailchimp/source_mailchimp/components.py

```diff
@@ -0,0 +1,120 @@
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+
+from dataclasses import dataclass
+from typing import Any, List, Mapping, Optional
+
+import pendulum
+import requests
+from airbyte_cdk.sources.declarative.auth.declarative_authenticator import DeclarativeAuthenticator
+from airbyte_cdk.sources.declarative.auth.token import BasicHttpAuthenticator, BearerAuthenticator
+from airbyte_cdk.sources.declarative.extractors.record_filter import RecordFilter
+from airbyte_cdk.sources.declarative.requesters.http_requester import HttpRequester
+from airbyte_cdk.sources.declarative.requesters.request_options.interpolated_request_options_provider import (
+    InterpolatedRequestOptionsProvider,
+    RequestInput,
+)
+from airbyte_cdk.sources.declarative.types import StreamSlice, StreamState
+
+
+@dataclass
+class MailChimpRequester(HttpRequester):
+    """
+    Defines `get_data_center_location` method to define data_center and update config on the fly, as it depends on Authenticator implementation.
+    """
+
+    request_body_json: Optional[RequestInput] = None
+    request_headers: Optional[RequestInput] = None
+    request_parameters: Optional[RequestInput] = None
+    request_body_data: Optional[RequestInput] = None
+
+    def __post_init__(self, parameters: Mapping[str, Any]) -> None:
+
+        self.request_options_provider = InterpolatedRequestOptionsProvider(
+            request_body_data=self.request_body_data,
+            request_body_json=self.request_body_json,
+            request_headers=self.request_headers,
+            request_parameters=self.request_parameters,
+            config=self.config,
+            parameters=parameters or {},
+        )
+        super().__post_init__(parameters)
+
+    def get_url_base(self) -> str:
+        self.get_data_center_location()
+        return super().get_url_base()
+
+    def get_data_center_location(self):
+        if not self.config.get("data_center"):
+            if isinstance(self.authenticator, BasicHttpAuthenticator):
+                data_center = self.config["credentials"]["apikey"].split("-").pop()
+            else:
+                data_center = self.get_oauth_data_center(self.config["credentials"]["access_token"])
+            self.config["data_center"] = data_center
```

Comment:
why are we modifying the config? As far as I know, this is an implementation detail. Why not remove the url_base field from the manifest?

Reply:
Mainly for `visibility` and not overriding the `get_url_base` method from base class

Reply:
I would agree with @girarda that it would be better to pass this as a parameter of `__init__` else we create an object that is stateful. That being said, I don't know how all of this fits in the initialization process of the custom components. The other solution I would propose is to not use custom components and use only interpolation